### PR TITLE
Add DutchV3 Order + Nonlinear decay

### PR DESCRIPTION
**Pull request by codyborn**
Add DutchV3 and NonlinearDecay logic to order.rs

### CODE REVIEW

File: crates/uniswapx-rs/src/order.rs [hunk 1]
```diff
@@ -239,7 +288,73 @@ impl PriorityOutput {
...
+
+    pub fn resolve(&self, block_number: u64, timestamp: u64) -> OrderResolution {
+        let timestamp = Uint::from(timestamp);
+
+        if self.info.deadline.lt(&timestamp) {
+            return OrderResolution::Expired;
+        };
+
+        // resolve over the decay curve
+        // TODO: apply cosigner logic
```
Comment (codyborn) [inline crates/uniswapx-rs/src/order.rs]: Pulled this from V2.. @zhongeric I see the cosigner overrides are being used and exclusivity is accounted for.. is there anything else that needs to be done?

File: crates/uniswapx-rs/src/order.rs [hunk 2]
```diff
@@ -285,6 +400,131 @@ fn resolve_decay(
...
+
+    /// Calculates the decayed amount based on the current block and the defined curve
+    pub fn decay(
+        &self,
+        start_amount: Uint<256, 4>,
+        decay_start_block: Uint<256, 4>,
+        current_block: Uint<256, 4>,
+        min_amount: Uint<256, 4>,
+        max_amount: Uint<256, 4>,
+    ) -> Result<Uint<256, 4>, Box<dyn Error>> {
```
Comment (codyborn) [inline crates/uniswapx-rs/src/order.rs]: `Box<dyn Error>` is this the best way to handle errors?

Comment (ConjunctiveNormalForm) [reply crates/uniswapx-rs/src/order.rs]: hmmm not sure if that's true... I think throughout this codebase we use specific error types and I don't see why we need to use trait objects here (and in other parts of this PR)

File: crates/uniswapx-rs/src/order.rs [hunk 3]
```diff
@@ -285,6 +400,131 @@ fn resolve_decay(
...
+            (current_block - decay_start_block).min(Uint::<256, 4>::from(u16::MAX))
+        )?;
+        println!("block_delta: {}", block_delta);
+
+        let (start_point, end_point, rel_start_amount, rel_end_amount) = 
+            self.locate_curve_position(block_delta)?;
+        println!("start_point: {}", start_point);
+        println!("end_point: {}", end_point);
+        println!("rel_start_amount: {}", rel_start_amount);
+        println!("rel_end_amount: {}", rel_end_amount);
```
Comment (ConjunctiveNormalForm) [inline crates/uniswapx-rs/src/order.rs]: I know you are just debugging, but tracing::info!/warn!/etc produce better formatted logs

File: crates/uniswapx-rs/src/order.rs [hunk 4]
```diff
@@ -285,6 +400,131 @@ fn resolve_decay(
...
+
+        // Calculate decay of only the relative amounts
+        let curve_delta = resolve_decay(
+            Uint::<256, 4>::from(block_delta),
+            Uint::<256, 4>::from(start_point),
+            Uint::<256, 4>::from(end_point),
+            Uint::<256, 4>::try_from(rel_start_amount).unwrap(),
+            Uint::<256, 4>::try_from(rel_end_amount).unwrap(),
```
Comment (ConjunctiveNormalForm) [inline crates/uniswapx-rs/src/order.rs]: I know we used `unwrap()` in in the existing code 😅 but generally should be avoided, e.g. can do `.expect("<what went wrong")` instead to make debugging easier

File: crates/uniswapx-rs/src/order.rs [hunk 5]
```diff
@@ -285,6 +400,131 @@ fn resolve_decay(
...
+            packed_data |= shifted_value;
+        }
+
+        Ok(packed_data)
+    }
+
+    
+    /// Retrieve the nth uint16 value from a packed uint256
+    fn get_element(packed_data: Uint<256, 4>, n: usize) -> Result<u16, Box<dyn std::error::Error>> {
+        if n >= 16 {
```
Comment (ConjunctiveNormalForm) [inline crates/uniswapx-rs/src/order.rs]: nit: 16 to constant

File: crates/uniswapx-rs/src/order.rs [hunk 6]
```diff
@@ -285,6 +400,131 @@ fn resolve_decay(
     }
 }
 
+impl NonlinearDutchDecay {
+
+    /// Calculates the decayed amount based on the current block and the defined curve
+    pub fn decay(
+        &self,
+        start_amount: Uint<256, 4>,
```
Comment (ConjunctiveNormalForm) [inline crates/uniswapx-rs/src/order.rs]: maybe cleaner to define type alias for Uint<256, 4>?

File: crates/uniswapx-rs/src/order.rs [hunk 7]
```diff
@@ -172,9 +221,9 @@ impl V2DutchOrder {
 
                 // add exclusivity override to amount
                 if self.cosignerData.decayStartTime.gt(&timestamp) && !self.cosignerData.exclusiveFiller.is_zero() {
-                    let exclusivity = self.cosignerData.exclusivityOverrideBps.wrapping_add(Uint::from(10000));
+                    let exclusivity = self.cosignerData.exclusivityOverrideBps.wrapping_add(BPS);
```
Comment (ConjunctiveNormalForm) [inline crates/uniswapx-rs/src/order.rs]: actually... is wrapping the proper behavior here? maybe better to do checked arithmetics and throw on error?

File: crates/uniswapx-rs/src/order.rs (hunk 3 above)
Comment (codyborn) [reply crates/uniswapx-rs/src/order.rs]: oops, didn't mean to include this

File: crates/uniswapx-rs/src/order.rs [hunk 8]
```diff
@@ -82,21 +83,69 @@ sol! {
...
+        uint256 startAmount;
+        NonlinearDutchDecay curve;
+        address recipient;
+        uint256 minAmount;
+        uint256 adjustmentPerGweiBaseFee;
+    }
 }
 
 pub const MPS: u64 = 1e7 as u64;
+pub const BPS: Uint<256, 4> = Uint::from_limbs([10000, 0, 0, 0]);
```
Comment (zhongeric) [inline crates/uniswapx-rs/src/order.rs]: why can't we just do u64 like above?

File: crates/uniswapx-rs/src/order.rs (hunk 7 above)
Comment (zhongeric) [reply crates/uniswapx-rs/src/order.rs]: yeah probably should throw

File: crates/uniswapx-rs/src/order.rs [hunk 9]
```diff
@@ -373,4 +613,268 @@ mod tests {
...
+        };
+
+        let start_block = Uint::from(1000);
+        let current_block = Uint::from(1100);
+        let start_amount = Uint::from(500); // Less than max relativeAmount
+        let min_amount = Uint::from(0);
+        let max_amount = Uint::MAX;
+
+        let result = decay.decay(start_amount, start_block, current_block, min_amount, max_amount);
+        assert!(result.is_err());
```
Comment (zhongeric) [inline crates/uniswapx-rs/src/order.rs]: shouldn't this be Ok() and bounded to 0 in the case of underflow?

File: crates/uniswapx-rs/src/order.rs (hunk 1 above)
Comment (zhongeric) [reply crates/uniswapx-rs/src/order.rs]: hmm seems fine to me

File: crates/uniswapx-rs/src/order.rs (hunk 2 above)
Comment (codyborn) [reply crates/uniswapx-rs/src/order.rs]: I see.. updated to use the implicit error response

File: crates/uniswapx-rs/src/order.rs (hunk 7 above)
Comment (codyborn) [reply crates/uniswapx-rs/src/order.rs]: Yeah looks like we do wrapped arithmetic all over but I added checks for some of it.

File: crates/uniswapx-rs/src/order.rs (hunk 8 above)
Comment (codyborn) [reply crates/uniswapx-rs/src/order.rs]: The thing it's being added to (exclusivityOverrideBps) is of type `Uint<256>`

File: crates/uniswapx-rs/src/order.rs (hunk 6 above)
Comment (codyborn) [reply crates/uniswapx-rs/src/order.rs]: Good idea

File: crates/uniswapx-rs/src/order.rs (hunk 9 above)
Comment (codyborn) [reply crates/uniswapx-rs/src/order.rs]: You are totally right.. thanks for paying attention.

File: crates/uniswapx-rs/src/order.rs [hunk 10]
```diff
@@ -6,6 +6,10 @@ use alloy_primitives::I256;
 use alloy_sol_types::sol;
 use anyhow::Result;
 
+use crate::sol_math::MulDiv;
+
+type U256 = Uint<256, 4>;
```
Comment (ConjunctiveNormalForm) [inline crates/uniswapx-rs/src/order.rs]: hmmm I recommend a more expressive alias name, especially given that U256 is a defined type in ethers already

or consider using Newtypes instead, e.g. 
```
pub struct TokenAmount(Uint<256,4>);
```

File: crates/uniswapx-rs/src/lib.rs [hunk 11]
```diff
@@ -1 +1,2 @@
 pub mod order;
+pub mod sol_math;
```
Comment (ConjunctiveNormalForm) [inline crates/uniswapx-rs/src/lib.rs]:
```suggestion
pub mod sol_math;

```

File: crates/uniswapx-rs/src/sol_math.rs [hunk 12]
```diff
@@ -0,0 +1,145 @@
+use alloy_primitives::Uint;
+
+type U256 = Uint<256, 4>;
+
+pub trait MulDiv {
+    fn mul_div_down(&self, b: U256, c: U256) -> Result<U256, anyhow::Error>;
+    fn mul_div_up(&self, b: U256, c: U256) -> Result<U256, anyhow::Error>;
+}
+
```
Comment (ConjunctiveNormalForm) [inline crates/uniswapx-rs/src/sol_math.rs:9]: I'm a little surprised this is not implemented in alloy/ethers

File: crates/uniswapx-rs/src/order.rs [hunk 13]
```diff
@@ -348,29 +690,402 @@ mod tests {
...
+                current_block,
+                min_amount,
+                max_amount,
+                *decay_func
+            );
+            assert!(result.is_ok());
+            assert_eq!(result.unwrap(), start_amount);
+        }
+    }
+}
```
Comment (ConjunctiveNormalForm) [inline crates/uniswapx-rs/src/order.rs]:
```suggestion
}

```

**Review by ConjunctiveNormalForm — APPROVED** [review]
lgtm, just a few nits

File: crates/uniswapx-rs/src/order.rs (hunk 10 above)
Comment (codyborn) [reply crates/uniswapx-rs/src/order.rs]: Updated to `BigUint`

**Pull request merged**